src/main/scala/sigmastate/serialization/DataSerializer.scala

```diff
@@ -38,31 +38,7 @@ object DataSerializer {
       val bytes = v.asInstanceOf[EcPointType].getEncoded(true)
       w.putBytes(bytes)
     case SBox =>
-      val obj = v.asInstanceOf[ErgoBox]
-      w.putLong(obj.value)
-      w.putBytes(obj.propositionBytes)
-
-      val nRegs = obj.additionalRegisters.keys.size
-      if (nRegs + ErgoBox.startingNonMandatoryIndex > 255)
-        sys.error(s"The number of non-mandatory indexes $nRegs exceeds ${255 - ErgoBox.startingNonMandatoryIndex} limit.")
-      w.put(nRegs.toByte)
-
-      // we assume non-mandatory indexes are densely packed from startingNonManadatoryIndex
-      // this convention allows to save 1 bite for each register
-      val startReg = ErgoBox.startingNonMandatoryIndex
-      val endReg = ErgoBox.startingNonMandatoryIndex + nRegs - 1
-      for (regId <- startReg to endReg) {
-        val reg = ErgoBox.findRegisterByIndex(regId.toByte).get
-        obj.get(reg) match {
-          case Some(v) =>
-            w.putValue(v)
-          case None =>
-            sys.error(s"Set of non-mandatory indexes is not densely packed: " +
-                      s"register R$regId is missing in the range [$startReg .. $endReg]")
-        }
-      }
-      w.putBytes(obj.transactionId)
-      w.putUShort(obj.index)
+      ErgoBox.serializer.serializeBody(v.asInstanceOf[ErgoBox], w)
```

Comment:
@greenhat If we do such code move, we essentially change what is primary, what is secondary.
Before, DataSerializer was primary place for data serialization. And Box.serializer reused it and was secondary (kind of wrapper). 
If we change this, we should accordingly change SGroupElement, SAvlTree etc. to be consistent.
I don't think we should do this at all, what are the reasons, am I missed something?

Reply:
I see your point. I saw `DataSerializer` as more of a place for "simple" types (de)serialization. As soon as logic gets complicated it deserves its own serializer. My only reason to extract box serialization code was to ease its understanding which should be better when it's isolated rather than in a `case` of a big switch.  I'm ready to extract `SGroupElement` and `SAvlTree` or move box serialization back to `DataSerializer`.

Reply:
@greenhat that is what I mean it should be done atomically, in all or nothing style, to be consistent.